include/ginkgo/core/log/profiler_hook.hpp

```diff
@@ -0,0 +1,246 @@
+    void on_criterion_check_started(const stop::Criterion* criterion,
+                                    const size_type& num_iterations,
+                                    const LinOp* residual,
+                                    const LinOp* residual_norm,
+                                    const LinOp* solution,
+                                    const uint8& stopping_id,
+                                    const bool& set_finalized) const override;
+
+    void on_criterion_check_completed(
+        const stop::Criterion* criterion, const size_type& num_iterations,
+        const LinOp* residual, const LinOp* residual_norm,
+        const LinOp* solution, const uint8& stopping_id,
+        const bool& set_finalized, const array<stopping_status>* status,
+        const bool& one_changed, const bool& all_stopped) const override;
+
+    void on_criterion_check_completed(
+        const stop::Criterion* criterion, const size_type& num_iterations,
+        const LinOp* residual, const LinOp* residual_norm,
+        const LinOp* implicit_sq_resnorm, const LinOp* solution,
+        const uint8& stopping_id, const bool& set_finalized,
+        const array<stopping_status>* status, const bool& one_changed,
+        const bool& all_stopped) const override;
+
+    bool needs_propagation() const override;
+
+    /**
+     * Sets the name for an object to be profiled. Every instance of that object
+     * in the profile will be replaced by the name instead of its runtime type.
+     * @param obj  the object
+     * @param name  its name
+     */
+    void set_object_name(const PolymorphicObject* obj, std::string name);
+
+    /**
+     * Should the events call executor->synchronize on operations and
+     * copy/allocation? This leads to a certain overhead, but makes the
+     * execution timeline of kernels synchronous.
+     */
+    void set_synchronization(bool synchronize);
+
+    /** The Ginkgo yellow background color as packed 32 bit ARGB value. */
+    constexpr static uint32 color_yellow_argb = 0xFFFFCB05U;
+
+    /**
+     * Creates a logger annotating Ginkgo events with TAU ranges via PerfStubs.
+     * Since TAU requires global initialization and finalization, this will only
+     * create a single global instance that is destroyed at program exit.
+     * @param initialize  Should we call TAU's initialization and finalization
+     *                    functions, or does the application take care of it?
+     */
+    static std::shared_ptr<ProfilerHook> create_tau(bool initialize = true);
+
+    /**
+     * Creates a logger annotating Ginkgo events with NVTX ranges for CUDA.
+     * @param color_argb  The color of the NVTX ranges in the NSight Systems
+     *                    output. It has to be a 32 bit packed ARGB value.
+     */
+    static std::shared_ptr<ProfilerHook> create_nvtx(
+        uint32 color_argb = color_yellow_argb);
+
+    /**
+     * Creates a logger annotating Ginkgo events with ROCTX ranges for ROCm.
```

Comment:
I think in the long run, we will not be able to maintain CUDA and ROCm support separately, also since HIP-CUDA doesn't play nice with CUDA 11 (or did that change in the last year?)

Reply:
My point is just that we don't make that distinction anywhere else. We don't reference rocm at all in our public headers. So, for users who just know hip as the equivalent of cuda for amd devices, now mentioning rocm could be confusing. And if we drop hip support for cuda (which seems reasonable, but I don't know enough of it) then there is no distinction at all, so we should pick either hip or rocm and use that one consistently.

Reply:
Fair point, CMake also calls it HIP despite only supporting the ROCm backend. I'll stick with HIP then.